Data/Primitive/Internal/Bit.hs

```diff
@@ -0,0 +1,63 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Data.Primitive.Internal.Bit
+ (
+   MutableBitArray
+ , newBitArray
+ , readBitArray
+ , setBitArray
+ ) where
+
+import Data.Primitive.ByteArray
+import Control.Monad.Primitive
+import Data.Bits
+
+newtype MutableBitArray s = MBA (MutableByteArray s)
+
+newBitArray :: PrimMonad m => Int -> m (MutableBitArray (PrimState m))
+newBitArray n = do
+  let s = ((n + wordSize - 1) `unsafeShiftR` 3)
+  mary <- newByteArray s
+  fillByteArray mary 0 s 0
+  return (MBA mary)
```

Comment:
Alternatively, we could refrain from filling the array here, and just use a `writeBitArray` function that takes the value to write. That's probably better, actually.